From b98961b4e37cded023bb68ca693fbdeb32990c44 Mon Sep 17 00:00:00 2001
From: Justin Tsai <justin@gmail.com>
Date: Wed, 10 May 2023 14:33:34 +0800
Subject: [PATCH 10/12]  add JIO partition and add reset to default sync mac to
 interface function

including:
1.get base mac and add 3 for 2g BSSID add 4 for 5g BSSID
openwrt/package/kernel/mac80211/files/lib/wifi/mac80211.sh
2.get base mac for wan mac and add 2 for br-lan mac
openwrt/target/linux/mediatek/mt7986/base-files/etc/board.d/02_network
3.add JIO partition in kernel device tree
openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986-spim-nand-partition.dtsi
4.add config lan 4 and kernel LED gpio for solid RED light when bootup to kernel.
openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-spim-nand-rfb.dts
---
 .../kernel/mac80211/files/lib/wifi/mac80211.sh     | 39 +++++++++++++++++++---
 .../dts/mediatek/mt7986-spim-nand-partition.dtsi   | 17 +++++++++-
 .../boot/dts/mediatek/mt7986a-spim-nand-rfb.dts    | 38 +++++++++++++++++++--
 .../mt7986/base-files/etc/board.d/02_network       | 15 ++++++---
 4 files changed, 96 insertions(+), 13 deletions(-)

diff --git a/openwrt/package/kernel/mac80211/files/lib/wifi/mac80211.sh b/openwrt/package/kernel/mac80211/files/lib/wifi/mac80211.sh
index 438bf92..9dbc5ec 100644
--- a/openwrt/package/kernel/mac80211/files/lib/wifi/mac80211.sh
+++ b/openwrt/package/kernel/mac80211/files/lib/wifi/mac80211.sh
@@ -153,6 +153,34 @@ check_board_phy() {
 	fi
 }
 
+
+gm_mfg_get_wifi_ssid_val()
+{
+	GM_WIFI_SSID_OFFSET=0x1d
+	GM_WIFI_SSID_MAX_LEN=15
+	
+	HEX_STR=`hexdump -v -n 15 -s 0x1d -e '14/1 "%02x-" "%02x "' /dev/mtd7 | sed 's/-/ \\\x/g'`
+	echo -e "\x${HEX_STR}" | sed 's/ //g'
+}
+
+gm_mfg_get_wifi_mac()
+{
+	interface="$1"
+	
+	wan_mac=$(hexdump -v -n 6 -s 0x00 -e '5/1 "%02x:" "%02x "' /dev/mtd7)
+	backup_wan=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${wan_mac})
+	lan_mac=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${backup_wan})
+	wifi_2g=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${lan_mac})
+	wifi_5g=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${wifi_2g})
+	
+	if [ "$interface" = "2g" ];then
+		echo "${wifi_2g}"
+	elif [ "$interface" = "5g" ];then
+		echo "${wifi_5g}"
+	fi
+
+}
+
 detect_mac80211() {
 	devidx=0
 	config_load wireless
@@ -200,7 +228,9 @@ detect_mac80211() {
 				dev_id="set wireless.${name}.phy='$dev'"
 				;;
 		esac
-
+		gm_factory_ssid="$(gm_mfg_get_wifi_ssid_val)"
+		gm_wifi_mac="$(gm_mfg_get_wifi_mac ${mode_band})"
+		
 		uci -q batch <<-EOF
 			set wireless.${name}=wifi-device
 			set wireless.${name}.type=mac80211
@@ -208,13 +238,14 @@ detect_mac80211() {
 			set wireless.${name}.channel=${channel}
 			set wireless.${name}.band=${mode_band}
 			set wireless.${name}.htmode=$htmode
-			set wireless.${name}.disabled=1
-
+			set wireless.${name}.disabled=0
+			
 			set wireless.default_${name}=wifi-iface
 			set wireless.default_${name}.device=${name}
 			set wireless.default_${name}.network=lan
 			set wireless.default_${name}.mode=ap
-			set wireless.default_${name}.ssid=OpenWrt
+			set wireless.default_${name}.macaddr='${gm_wifi_mac}'
+			set wireless.default_${name}.ssid=${gm_factory_ssid}_${mode_band}
 			set wireless.default_${name}.encryption=${encryption}
 
 EOF
diff --git a/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986-spim-nand-partition.dtsi b/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986-spim-nand-partition.dtsi
index 4cc7961..48f1131 100644
--- a/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986-spim-nand-partition.dtsi
+++ b/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986-spim-nand-partition.dtsi
@@ -36,7 +36,22 @@
 
 			partition@580000 {
 				label = "ubi";
-				reg = <0x580000 0x4000000>;
+				reg = <0x580000 0x8c00000>;
+			};
+			
+			partition@9180000 {
+				label = "ubi2";
+				reg = <0x9180000 0x5a00000>;
+			};
+			
+			partition@eb80000 {
+				label = "mfg";
+				reg = <0xeb80000 0x0200000>;
+			};
+			
+			partition@ed80000 {
+				label = "Jio-Reserved";
+				reg = <0xed80000 0x0200000>;
 			};
 		};
 	};
diff --git a/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-spim-nand-rfb.dts b/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-spim-nand-rfb.dts
index 9c06c6c..3255dbc 100644
--- a/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-spim-nand-rfb.dts
+++ b/openwrt/target/linux/mediatek/files-5.4/arch/arm64/boot/dts/mediatek/mt7986a-spim-nand-rfb.dts
@@ -2,6 +2,9 @@
 #include "mt7986a.dtsi"
 #include "mt7986a-pinctrl.dtsi"
 #include "mt7986-spim-nand-partition.dtsi"
+#include <dt-bindings/input/linux-event-codes.h>
+#include <dt-bindings/gpio/gpio.h>
+
 / {
 	model = "MediaTek MT7986a RFB";
 	compatible = "mediatek,mt7986a-spim-snand-rfb";
@@ -31,6 +34,35 @@
 		mediatek,ext-codec = <&proslic_spi>;
 		status = "okay";
 	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 9 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		led_r {
+			label = "led_r";
+			gpios = <&pio 12 GPIO_ACTIVE_LOW>;
+			default-state = "keep";
+		};
+		led_g {
+			label = "led_g";
+			gpios = <&pio 13 GPIO_ACTIVE_LOW>;
+			default-state = "keep";
+		};
+		led_b {
+			label = "led_b";
+			gpios = <&pio 14 GPIO_ACTIVE_LOW>;
+			default-state = "keep";
+		};
+	};
+
 };
 
 &fan {
@@ -134,9 +166,9 @@
 				#address-cells = <1>;
 				#size-cells = <0>;
 
-				port@0 {
-					reg = <0>;
-					label = "lan0";
+				port@4 {
+					reg = <4>;
+					label = "lan4";
 				};
 
 				port@1 {
diff --git a/openwrt/target/linux/mediatek/mt7986/base-files/etc/board.d/02_network b/openwrt/target/linux/mediatek/mt7986/base-files/etc/board.d/02_network
index 0b61e50..1a46821 100755
--- a/openwrt/target/linux/mediatek/mt7986/base-files/etc/board.d/02_network
+++ b/openwrt/target/linux/mediatek/mt7986/base-files/etc/board.d/02_network
@@ -28,7 +28,8 @@ mediatek_setup_interfaces()
 mediatek_setup_macs()
 {
 	local board="$1"
-	local part_name="Factory"
+	#local part_name="Factory"
+	local part_name="mfg"
 	local lan_mac=""
 	local wan_mac=""
 	local lan_mac_offset=""
@@ -36,14 +37,18 @@ mediatek_setup_macs()
 
 	case $board in
 	*)
-		lan_mac_offset="0x2A"
-		wan_mac_offset="0x24"
+		#lan_mac_offset="0x2A"
+		#wan_mac_offset="0x24"
+		#lan_mac_offset="0x06"
+		wan_mac_offset="0x00"
 		;;
 	esac
 
-	lan_mac=$(mtd_get_mac_binary $part_name $lan_mac_offset)
+	#lan_mac=$(mtd_get_mac_binary $part_name $lan_mac_offset)
 	wan_mac=$(mtd_get_mac_binary $part_name $wan_mac_offset)
-
+	backup_wan=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${wan_mac})
+	lan_mac=$(/etc/mfg/gm_mtc/increment_mac_by_1.sh ${backup_wan})
+	
 	case "$lan_mac" in
 		00:00:00:00:00:00);;
 		ff:ff:ff:ff:ff:ff);;
-- 
2.7.4

